File: 1201-1300/1298-Maximum Candies You Can Get from Boxes/1298-Maximum Candies You Can Get from Boxes.cpp
class Solution {
public:
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        queue<int> Q;
        unordered_set<int> possible;
        for (int box : initialBoxes) {
            if (status[box] == 1) {
                Q.push(box);
            }
            else {
                possible.insert(box);
            }
        }
        
        int num = 0;
        while (!Q.empty()) {
            int box = Q.front();
            Q.pop();
            num += candies[box];
            for (int key : keys[box]) {
                status[key] = 1;
                if (possible.count(key)) {
                    possible.erase(key);
                    Q.push(key);
                }
            }
            
            for (int cb : containedBoxes[box]) {
                if (status[cb] == 1) {
                    Q.push(cb);
                }
                else {
                    possible.insert(cb);
                }
            }
        }
        
        return num;
    }
};

File: 1201-1300/1298-Maximum Candies You Can Get from Boxes/1298-Maximum Candies You Can Get from Boxes.py
import collections
class Solution:
    def maxCandies(self, status: List[int], candies: List[int], keys: List[List[int]], containedBoxes: List[List[int]], initialBoxes: List[int]) -> int:
        num = 0
        Q = collections.deque([box for box in initialBoxes if status[box] == 1])
        possible = set([box for box in initialBoxes if status[box] == 0])
        while Q:
            box = Q.popleft()
            num += candies[box]
            for key in keys[box]:
                status[key] = 1
                if key in possible:
                    possible.remove(key)
                    Q.append(key)
            
            for cb in containedBoxes[box]:
                if status[cb] == 1:
                    Q.append(cb)
                else:
                    possible.add(cb)

        return num
